src/garage/sampler/worker.py

```diff
@@ -0,0 +1,277 @@
+    Args:
+        seed(int): The seed to use to intialize random number generators.
+        max_path_length(int or float): The maximum length paths which will
+            be sampled. Can be (floating point) infinity.
+        worker_number(int): The number of the worker where this update is
+            occurring. This argument is used to set a different seed for each
+            worker.
+
+    Attributes:
+        agent(Policy or None): The worker's agent.
+        env(gym.Env or None): The worker's environment.
+
+    """
+
+    def __init__(
+            self,
+            *,  # Require passing by keyword, since everything's an int.
+            seed,
+            max_path_length,
+            worker_number):
+        super().__init__(seed=seed,
+                         max_path_length=max_path_length,
+                         worker_number=worker_number)
+        self.agent = None
+        self.env = None
+        self._observations = []
+        self._actions = []
+        self._rewards = []
+        self._terminals = []
+        self._lengths = []
+        self._agent_infos = defaultdict(list)
+        self._env_infos = defaultdict(list)
+        self._last_obs = None
+        self._path_length = 0
+        self.worker_init()
+
+    def worker_init(self):
+        """Initialize a worker."""
+        deterministic.set_seed(self._seed + self._worker_number)
+
+    def update_agent(self, agent_update):
+        """Update an agent, assuming it implements garage.Policy.
+
+        Args:
+            agent_update(Dict[str, np.array]): Parameters to agent, which
+                should have been generated by calling
+                `policy.get_param_values`. Note that other implementations of
+                `Worker` may take different types for this parameter.
+
+        """
+        self.agent.set_param_values(agent_update)
+
+    def update_env(self, env_update):
+        """Use any non-None env_update as a new environment.
+
+        A simple env update function. If env_update is not None, it should be
+        the complete new environment.
+
+        This allows changing environments by passing the new environment as
+        `env_update` into `obtain_samples`.
+
+        Args:
+            env_update(gym.Env or None): The environment to replace the
+                existing env with. Note that other implementations of
+                `Worker` may take different types for this parameter.
+
+        """
+        if env_update is not None:
+            self.env.close()
+            self.env = env_update
+
+    def start_rollout(self):
+        """Begin a new rollout."""
+        self._observations = []
+        self._actions = []
+        self._rewards = []
+        self._terminals = []
+        self._lengths = []
+        self._agent_infos = defaultdict(list)
+        self._env_infos = defaultdict(list)
+        self._path_length = 0
+        self._last_obs = self.env.reset()
+        self.agent.reset()
+
+    def step_rollout(self):
```

Comment:
should we be using the gym termination signal for keeping track of how environment termination occurs. Ironically enough, sac implementation in spinning up keeps track of environment termination this way as well, even though open ai created gym, and the environment wrapper that solves this problem.

Reply:
I think we should fix this universally throughout garage (probably in `GarageEnv`) rather than implementing a fix in every sampler.